### DISCUSSION TITLE
For some reason in some circumstances this code doesn't work as expected

### DISCUSSION THREAD

**Discussion by RIP21** [urql-graphql/urql#3282] [Q&A]
<img width="644" alt="CleanShot 2023-06-19 at 10 55 08@2x" src="https://github.com/urql-graphql/urql/assets/3940079/f206ac13-9fa4-4d60-8671-65e26e89b14a">
For some reason, safe me query is not getting re-executed.
Organization me query always gets re-executed.

Insanely weird behaviour. If I login, and then trigger the code that calls these lines on the screenshot, it is not working.
If I already logged in and refresh the page and do exactly the same on the same page (urql cache state is exactly the same as after login, from logged out page and getting redirected to that same page) it doesn't.

It's happening in Next.js and it's happening in non-suspended mode (suspense: false) as suspense is broken with URQL and Next.js in production and had to disable it as if Next.js suspended routes are conflicting with URQL suspension. (old mode, not app dir)

Don't know whom to blame, but URQL feels VERY unstable. At least in v4 :( 

It's with graphcache enabled, and in dev mode with devtools extension.

I'll test a production build this evening (CEST time) and see how it behaves on production, but yeah, super annoying...

**Comment by kitten** [comment]
I don't really have enough details to give a clear answer here, but reexecute calls are inherently **UI concerns**

Those are designed to behave more heuristically. Instead, it seems to me like you're trying to do two **imperative** things at once.I could be wrong, but have you checked whether the `push` is unmounting the query that isn't being reexecuted?

Maybe add a `debugExchange` and check for teardowns for the relevant query

**Reply by RIP21** [reply]
Here are logs of the operation that is struggling to reexecute.
It's how it looks at the page load.
<img width="892" alt="CleanShot 2023-06-19 at 22 18 25@2x" src="https://github.com/urql-graphql/urql/assets/3940079/4389a5fa-4669-4f17-8250-0cfc88aaf5ba">

It fails to execute it even when I try to run it with `client.query(Document).toPromise()` in that case Promise is stuck forever and redirect is not happening (e.g. `push`) and I cannot see any Incoming operation logged via debug exchange despite `console.log` before it clearly shows that code was triggered.

Then when I log in it adds this to the log (two logs after the Complete operation, one is shown in screenshot, one is the same as the failed Completed operation before, was filtered in the screenshot as I haven't uncollapsed it so key is visible so it's shown)
<img width="889" alt="CleanShot 2023-06-19 at 22 22 30@2x" src="https://github.com/urql-graphql/urql/assets/3940079/74f75bfa-ab03-415a-a7d8-8a7d7af72413">

Then clicking anywhere that triggers code that supposedly reexecutes it using react useQuery APIs or client.query.toPromise results in no operation being triggered/no logs.

So there is no teardown it seems, plus in devtools extension, I can also see that the operation is still ongoing/listened to.

If I refresh the page, I see this operation with exactly the same key, and it looks exactly the same with the last status being "query", but when I trigger the code that rexecutes the operation it gets triggered as expected logging a bunch of "query" logs.

And in both, when it works (when I refresh the page being already logged in) and while not logged in (when it's not working), it starts with a bunch of `query, teardown, query, teardown, query, teardown`, and then just `query, query, query, query, query` (as this particular query used in every route (as it's used in `_app`))

**Reply by kitten** [reply]
I reckon it _could_ be Graphcache's looping protection, but it's impossible for me to tell.

I've been meaning to make some changes to that to make it less likely to trigger. To be frank, it triggers in weird scenarios that we haven't in the past taken into account, like a self-repeating query. It's basically designed to prevent multiple cache misses (especially in scenarios where the query has errored) from triggering each other.

This actually would be plausible to me. No matter whether you're running one or two queries, you have a query that by definition (i.e. having errored) will always experience a cache miss, and you're trying to basically force it to continue, which Graphcache by design stops.

The problem here is that we haven't found a good stopping condition — i.e. it's hard for us to tell which operations to stop and which to let through, given that there's many ways — including in UI code — to cause looping requests.
In this case, it's legitimate, but that's hard to detect.

Anyway, you should be able to confirm this by checking whether this condition (`!reexecutingOperations.has(res.operation.key)`) is evaluated and becomes `false`, i.e. if `reexecutingOperations` contains your key.
https://github.com/urql-graphql/urql/blob/6a441884790bf6b07acb553f7bdc3702c3a1c315/exchanges/graphcache/src/cacheExchange.ts#L333-L341

That said, unfortunately, I have little hopes of this being the issue, since `network-only` usually specifically bypasses this logic, as per: https://github.com/urql-graphql/urql/blob/6a441884790bf6b07acb553f7bdc3702c3a1c315/exchanges/graphcache/src/cacheExchange.ts#L314-L330

But this is currently presenting with too many unknowns for me to go for any specifics.

That said, what I do recommend if you're just looking for a resolution (which is mentioned in a couple of discussions/issues as the recommended approach) is to recreate the `Client` when a user logs in or out. This also ensures that your cache is erased and no information leaks across.

I assume you've got a good reason though for not doing this, and of course, this should still work.

That said, the only notes I can leave are:
- make sure that `@urql/core` is up-to-date and **not duplicated** (i.e. ensure the version you specify is actually the one you're using 😅 importing from `@urql/core` does rule this out, although it's a good idea to deduplicate it)
- let's make sure that Graphcache is up-to-date as well, if it's not. `6.1.0` specifically addressed a `reexecutingOperations` problem
- Make sure to continue using `client.query(Document, {/*vars*/}, { requestPolicy: 'network-only' }).toPromise()` in testing for now (no need to await it though btw), as it bypasses _a lot_ of code I can't see, and can't take into account when guessing where the problem may be

**Reply by RIP21** [reply]
Your guessing about queries that errored out is correct.
This query errors out if a user is not logged in into "safe" (one of our models), after we login, we re-execute that query and it should be successful in that case populating `safeMe` query with data. Before that, it's indeed 5-15 cache misses in case user is not logged in to "safe".

I also tried that in the production build, and in that case, I see that query gets triggered (in the network tab), but then the tab freezes completely.

So, I think you're guessing here is VERY close to the problem. I just need to understand how to work around it.

I'm SO SO SO SO SO SO SO tired of that :) I have a long-opened branch with 270 files affected and I'm so tired of merging conflicts :) At first, it was a partial rewrite from Apollo, to URQL with Suspense, then I got into weird behaviour (one of the issues I have opened now still hanging, this one: https://github.com/urql-graphql/urql/issues/3254), which was not major so we opt to ignore it. But then we've got, probably, exactly the same issue on production, that I've got now with the tab getting frozen on safe login, hence, the SAME. I thought it was due to Suspense (Discussion: https://github.com/urql-graphql/urql/discussions/3262), so I've disabled `suspense: true`, refactored a bunch of code again aaaaaaand got stuck with the same issue, but now it reproduces on development (tab is not stuck, but the operation is not triggered, while in production it freezes the tab)

So, yeah, I'm so tired of that, but it seems that we're closer to the solution...

Having all this info, @kitten can you suggest me a bunch of steps that will help me to debug it, or just what NOT to do to avoid it?
Because one thing that I already learned, is that URQL may get crazy in case one would abstract one global query that is very commonly used into some `useSafe` that will be used in 20 places on the page.

E.g. 
```tsx
/*
 * This hook has to be used in case when you would want to have a logged in safe data, but you're fine if it's not there (e.g. you're not logged in into safe)
 */
export const useNullableSafe = () => {
  return useSafeMeQuery()[0].data?.safeMe
}

export const useSafeMeQuery = () => {
  return useQuery({
    query: UseSafe_GetSafeDocument,
  })
}

/*
 * This hook is contrary to useNullableSafe, hence if you use it, you have to be sure that user is logged in into safe and safe is already fetched
 * otherwise it will throw an error
 */
export const useSafe = () => {
  const [{ data }] = useQueryOrThrow({
    query: UseSafe_GetSafeDocument,
  })
  if (!data?.safeMe) {
    throw new Error(
      "Safe is not defined, you need to guard this hook, if you haven't fetched it yet with useSafeMeQuery",
    )
  }
  return data.safeMe
}
```

Which is basically meaning reusing the same Document 15-20 times on the page. E.g. calling 
```ts
 useQuery({
    query: UseSafe_GetSafeDocument,
  })
  ```
  in 20 components on the same page.
  
  I thought that used-to-be-dedupe-exchange and now default client, should help with that case and it's OK, but it seems it may cause some weird things to happen. Probably, react implementation is to blame, but maybe not.

**Reply by kitten** [reply]
> I thought that used-to-be-dedupe-exchange and now default client, should help with that case and it's OK, but it seems it may cause some weird things to happen. Probably, react implementation is to blame, but maybe not.

It should be fine, tbf. Not that it's awesome to do tbh, but I can understand how it's easy to fall into that pattern.

For what it's worth, I don't think that should stop it from working but we'll see.

On your last `Incoming Operation`, have you validated that the `requestPolicy` is indeed set to `network-only`?

Other than that, re. debugging, tbh, it's an unexpected issue, and given that there's no output, the only thing that _can_ be done realistically is tracking down the issue by chance (which is unlikely, as there's only so many guesses I can make, and what I said above doesn't apply to `network-only`)
or to modify the `core` and `graphcache` code until the issue is caught, or stepping through it in a debugger.

> I also tried that in the production build, and in that case, I see that query gets triggered (in the network tab), but then the tab freezes completely.

So you can see a request being fired?
That does mean that the network result is successful.

I also don't really understand why the tab would be freezing (also not sure which kind of freeze you mean; because it can be three different kinds, I suppose? i.e. no interactivity, browser hangs due to a synchronous busy loop, browser slow down due to async busy loop)

Not really joking, but if you do get a request, and we can validate that a result enters Graphcache, I have no idea what could cause this, as that's really weird.

The only way to verify that though (without the debugging steps above), is to _move_ the `debugExchange` around and to see whether you can see a result anywhere in the exchange chain, and where you don't.
Assuming the network request does start, you'll always see the operation, but the result should get lost somewhere

**Reply by RIP21** [reply]
A little update. Once I stopped awaiting for `client.query.toPromise` it started to work for both development and production builds.

But if I use reexecute with network only (as in the very first example I gave in the first post) that I'm getting from `useQuery` it's not working in case 
`refresh on login page => login => redirect to user page => login into safe (no incoming operation triggered) `.
And works in the case 
`already logged in (session cookie) =>  refresh on the user page => login into safe (works as expected, new incoming operation, API call and redirect to safe page as a result)`

**Reply by RIP21** [reply]
> On your last Incoming Operation, have you validated that the requestPolicy is indeed set to network-only?

Absolutely

> I also don't really understand why the tab would be freezing (also not sure which kind of freeze you mean; because it can be three different kinds, I suppose? i.e. no interactivity, browser hangs due to a synchronous busy loop, browser slow down due to async busy loop)

Devtools of the browser are responsive, but no preview, or response info is seen in the network tab, same with the console, that kinda works, but doesn't really, so devtools of the browser are kinda alive, but the cursor and other things on the page are stuck and no hover animations and other things are working. So it's probably async kind of freeze. And with suspense issue mentioned above it's like that too.

It's important to mention that when I was doing `client.query.toPromise` with `await` in `useEffect` in NO suspense mode it started to freeze exactly the same way as the problem mentioned above e.g. https://github.com/urql-graphql/urql/discussions/3262 
Removing `await` made it work just fine in non-suspended scenario (unsure if that would help with `suspense: true` too as I refactored suspense stuff out, but even if it fixes it, still feels like a hack if fixes, so it's rather a bug, than a feature)

**Reply by kitten** [reply]
Yea, it'd make more sense that `client.query` works. As mentioned though, that doesn't get us very far if we don't know whether Graphcache updates the cache 😅

Removing the `await` allowing you to proceed only confirms that the `Client` isn't receiving the result

**Reply by RIP21** [reply]
toPromise call actually updates the cache as UI proceeds to rerender and show expected result.

So, TBH I have no idea why rexecute from useQuery behaves any differently (I haven't looked into its implementation recently, but I expect it to be pretty much the same thing, maybe without toPromise (probably in suspense mode it uses promise tho).

**Reply by kitten** [reply]
> probably in suspense mode it uses promise tho

React Suspense does use promises, but the implementation is entirely different, as the characeristics of a suspending component is different.

> I have no idea why rexecute from useQuery behaves any differently

I think something is missing here still. It's practically not possible for `executeQuery` not to work here, unless you're unmounting the `useQuery` hook's component or interrupting the request otherwise due to the `push`, as I've previously theorised.

> toPromise call actually updates the cache as UI proceeds to rerender and show expected result.

This makes me think that the result is somehow getting lost. However, this can only happen when the operation is triggered as the result comes in and `toPromise` receives it.

Basically, I've got the debugging steps above for you, but with the current information, this isn't really making sense, and my biggest suspicion is that the logs above, or some other description of this isn't accurate. Sorry! 😅

It's hard for me to reconcile the described behaviour here without details. I'm not denying that there may be a bug in here. But the problem is that the descriptions here of the behaviour don't add up for me, and additionally, I'm lacking enough information. On top of that, I still can't rule out that there's additional meddling in either UI code (or otherwise) that may changing or introducing some of the buggy behaviour here

**Reply by RIP21** [reply]
@kitten once I'm back from vacation I'll try my best to reread it all and try to narrow it down. If no clear indicators, then, well, I'm always open to a screen share call to try to boil it down in real time.

It's just, I love URQL (especially since you're insanely responsive usually), but hate in the same time, as its reactive nature makes it quite hard to adopt if you do that on something already quite big. If you use it from the day one, it's rather smooth ride.

Another question, is there a unit test or e2e test somewhere in the codebase of urql where I can see all possible state transitions of a query?

For instance, failed query with error, apparently may have a stale to be true, despite operation is finished and not in progress of refetching (network call finished already). I expect fetching to be mostly first load and stale to be reexecute in procees. Either failed or successful query. In case of failed, I don't expect it to ever be stale, as error response is always a cache miss so fetching should be getting into true in case of reexecution up until first successful result comes in. 
Afterwards tho, it will depend on what kind of rexecute you do, it it's network-only I assume it invalidates successful result?

Having a table or sort of state machine of all the possible states would be super helpful. 

Simplified, like, no partial response is possible.(tbf all of the queries and mutations we do are 1 query and never multiple, and aren't streamed etc, and I expect 98% of urql users are the same, so lots of the capabilites of urql and so the possible states are not relevant for us) 

And complicated one, with defer, live, partial results etc.

**Reply by kitten** [reply]
> Another question, is there a unit test or e2e test somewhere in the codebase of urql where I can see all possible state transitions of a query?
> Having a table or sort of state machine of all the possible states would be super helpful.

I hear this a lot, but to put it short, every single state is possible, and every single transition is possible:
- data can be set after a result
- fetching may or may not become true; depends on if:
  - how bindings choose to implement and update it, as it's a UI/binding-only flag
  - if the result is cached i.e. available synchronously
- stale can be:
  - initially true if a network request will be started but request is ongoing
  - can become true if an operation starts for the same result
- error & data:
  - can be independently set to value or `null`
  - `data` can of course be partial, if `graphQLErrors` are present
  - `data` is allowed to be present even if any error is set, to allow exchanges / custom exchanges to pass it
- `operations` is guaranteed if `fetching: false` tbf, but that's a detail that's not worth typing
- any value may change or persist during `fetching: true`/`stale: true` as bindings will update selectively and keep their last state alive
  - again: this is up to bindings

i.e. any state transition and combination is legal and has a corresponding use-case.

> I expect fetching to be mostly first load and stale to be reexecute in procees. 

That's basially the case — except, bindings provide `fetching`, but for all intents and purposes that's the same, except there's no concept of "initial" and I'd recommend not to rely on that assumption too much. `fetching` will and can become `true` whenever inputs change and a new operations result source is used.

> For instance, failed query with error, apparently may have a stale to be true, despite operation is finished and not in progress of refetching (network call finished already).

Yes, but `stale` just tells you that there's new data coming. It doesn't mean you have to ignore the error, unless it concerns your UI.

> In case of failed, I don't expect it to ever be stale

Again, `stale` just says a new operation is in progress. That doesn't mean that failures stop requests. This is because it's not illegal for an exchange to cache errors.

> Afterwards tho, it will depend on what kind of rexecute you do, it it's network-only I assume it invalidates successful result?

Yes, issuing `network-only` is defined to skip any cache. This means that it'll always result in an immediate request and your bindings will receive a `stale: true` update in the meantime.

> Simplified, like, no partial response is possible.

This is an easy trap in GraphQL to fall into.
Partial results are _important_ and by definition GraphQL can always deliver partial results at nullable fields 😅 

> and so the possible states are not relevant for us

The beauty of all of these states is that you can ignore them as needed though. `stale` is irrelevant if you don't use it. It's by design safe to ignore, if you know you don't need to know whether a new result is coming.

Similarly, you don't have to handle partial results. If you for instance use TypeScript's typegen, in most cases, you'd of course choose to let the UI handle and interpret nullish values without accounting for errors.

**Reply by RIP21** [reply]
Thanks for such a good response. 
> This is an easy trap in GraphQL to fall into.
Partial results are important and by definition GraphQL can always deliver partial results at nullable fields 😅

I know I know :) I'm talking mostly that we have only one root query, so it's not that we're going to handle multiple of them quite often. I know all the nuances of GraphQL, just don't want to overcomplicate things :) But it seems it's not that simple if it comes to GQL clients :P

**Comment by RIP21** [comment]
> toPromise call actually updates the cache as UI proceeds to rerender and show expected result.

This is stopped working. Lol.

**Reply by RIP21** [reply]
How do bindings behave when a client gets changed (recreated)? Will they set fetching or stale while "refetching"?

**Reply by RIP21** [reply]
Aaaand, nothing works at this point. Killing (recreating) the urql client on login to safe also freezes the tab in the production build...

I guess I'll have to rewrite everything back to Apollo and name it a day.

**Reply by frederikhors** [reply]
Can you please create a minimum reproduction to further investigate? @RIP21 ?

**Reply by RIP21** [reply]
> Can you please create a minimum reproduction to further investigate? @RIP21 ?

Sadly that would take more time than rewriting it to Apollo. And I doubt I'll manage to reproduce that on a smaller case.

So, I would rather refactor it back to Apollo, and then, after we redesign our GraphQL in a way, there is no need to reset cache as much, so it's needed only on logout, I'll probably try to use URQL again. Unless we move to Redux Toolkit Query and name it a day as it's stupid and simple and ultra reliable :D

Thanks for all the help guys, but I need to ship stuff now ASAP, it already took a 2 weeks extra due to this bug/problem/misuse, which messed up lots of things, and I cannot try to chew through it any longer.
